core/src/main/scala/org/apache/spark/deploy/master/Master.scala

```diff
@@ -796,9 +796,12 @@ private[deploy] class Master(
   }
 
   private def relaunchDriver(driver: DriverInfo) {
-    driver.worker = None
-    driver.state = DriverState.RELAUNCHING
-    waitingDrivers += driver
+    removeDriver(driver.id, DriverState.RELAUNCHING, None)
+    val newDriver = createDriver(driver.desc)
```

Comment:
Do you have a good reason to remove and create the driver in this case? It looks like some kind of overkill compared to the old logic.